mono/eglib/giconv.c

```diff
@@ -1304,3 +1304,49 @@ g_ucs4_to_utf16 (const gunichar *str, glong len, glong *items_read, glong *items
 	
 	return outbuf;
 }
+
+gchar*
+g_wcs_to_utf8 (const wchar_t *str, glong length)
+// "wcs" is what ANSI C and C++ call "wide character string".
+// For any "str" -- strlen, strchr, strtok, etc. there is "wcs" -- wcslen, wcschr, wcstok, etc.
+// where char has been replaced by wchar_t in the signature.
+// wchar_t is typically 16 bits on Windows and 32 bits on Unix.
+// $ echo "char a = sizeof(wchar_t);" | gcc -xc++ - -S -o - | grep [0-9]
+//	.byte	4                       ## 0x4
+{
+	g_assert (sizeof (wchar_t) == 1 || sizeof (wchar_t) == 2 || sizeof (wchar_t) == 4);
+	gchar* result;
+
+	if (!str)
+		return NULL;
+
+	switch (sizeof (wchar_t)) {
```

Comment:
This is wrong. Not only is the size of `wchar_t` implementation-defined, but so is the encoding. Even if you know that `sizeof(wchar_t) == 4` just casting it to `gunichar` doesn't make it a unicode character.